onnx/defs/data_type_utils.cc

```diff
@@ -8,404 +8,163 @@
+    case onnx::TensorProto_DataType_UINT8:
+      return "uint8";
+    case onnx::TensorProto_DataType_INT8:
+      return "int8";
+    case onnx::TensorProto_DataType_UINT16:
+      return "uint16";
+    case onnx::TensorProto_DataType_INT16:
+      return "int16";
+    case onnx::TensorProto_DataType_INT32:
+      return "int32";
+    case onnx::TensorProto_DataType_INT64:
+      return "int64";
+    case onnx::TensorProto_DataType_STRING:
+      return "string";
+    case onnx::TensorProto_DataType_BOOL:
+      return "bool";
+    case onnx::TensorProto_DataType_FLOAT16:
+      return "float16";
+    case onnx::TensorProto_DataType_DOUBLE:
+      return "double";
+    case onnx::TensorProto_DataType_UINT32:
+      return "uint32";
+    case onnx::TensorProto_DataType_UINT64:
+      return "uint64";
+    case onnx::TensorProto_DataType_COMPLEX64:
+      return "complex64";
+    case onnx::TensorProto_DataType_COMPLEX128:
+      return "complex128";
+    default:
+      assert(false);
+      return "";
   }
 }
 } // namespace Utils
+
+template <typename T>
+PDataType Abstract<T>::Type(
+    const std::string& name,
+    const std::string& domain) {
+  static Abstract abs_type(name, domain);
+  return &abs_type;
+}
+
+template <int elemT>
+PDataType TensorType<elemT>::Type() {
+  static TensorType tensor_type;
+  return &tensor_type;
+}
+
+const std::string& DataType::Domain() const {
+  return domain;
+}
+const std::string& DataType::Id() const {
+  return id;
+}
+const std::string& DataType::Description() const {
+  return description;
+}
+const TypeProto& DataType::ToProto() const {
+  return type_proto;
+}
+
+PDataType DataType::Tensor_FLOAT =
+    TensorType<TensorProto_DataType_FLOAT>::Type();
+PDataType DataType::Tensor_UINT8 =
+    TensorType<TensorProto_DataType_UINT8>::Type();
+PDataType DataType::Tensor_INT8 = TensorType<TensorProto_DataType_INT8>::Type();
+PDataType DataType::Tensor_UINT16 =
+    TensorType<TensorProto_DataType_UINT16>::Type();
+PDataType DataType::Tensor_INT16 =
+    TensorType<TensorProto_DataType_INT16>::Type();
+PDataType DataType::Tensor_INT32 =
+    TensorType<TensorProto_DataType_INT32>::Type();
+PDataType DataType::Tensor_INT64 =
+    TensorType<TensorProto_DataType_INT64>::Type();
+PDataType DataType::Tensor_STRING =
+    TensorType<TensorProto_DataType_STRING>::Type();
+PDataType DataType::Tensor_BOOL = TensorType<TensorProto_DataType_BOOL>::Type();
+PDataType DataType::Tensor_FLOAT16 =
+    TensorType<TensorProto_DataType_FLOAT16>::Type();
+PDataType DataType::Tensor_DOUBLE =
+    TensorType<TensorProto_DataType_DOUBLE>::Type();
+PDataType DataType::Tensor_UINT32 =
+    TensorType<TensorProto_DataType_UINT32>::Type();
+PDataType DataType::Tensor_UINT64 =
+    TensorType<TensorProto_DataType_UINT64>::Type();
+PDataType DataType::Tensor_COMPLEX64 =
+    TensorType<TensorProto_DataType_COMPLEX64>::Type();
+PDataType DataType::Tensor_COMPLEX128 =
+    TensorType<TensorProto_DataType_COMPLEX128>::Type();
+
```

Comment:
Do we also need to add DataType for "Seq"?

Reply:
@houseroad no, abstract type now is behaving as an named opaque type. We're not moving in the direction of defining all static types.

Reply:
I see. We don't have any operator which uses "Seq", yeah, no need to add it here.